[Shamanbenny] iP

Shamanbenny opened the pull request:
# NightCoder: A Playful and Motivational Chatbot for Late-Night Coding Sessions

## Features Implemented
- Interactive chatbot that assists with managing coding tasks.
- Supports task management operations such as:
  - Adding, listing, marking, and deleting tasks.
  - Supports **To-Do**, **Deadline**, and **Event** tasks.
  - **Persistent** task storage via file-based system.
  - ~~FAST~~ *SUPER FAST* to use

## Code Structure
1. **`NightCoder.java`** – Entry point of the application.
2. **`Ui.java`** – Handles all user interactions and displays messages.
3. **`Parser.java`** – Interprets and processes user commands.
4. **`Task.java`** – Abstract representation of a task.
5. **`TaskList.java`** – Manages list operations for tasks.
6. **`Storage.java`** – Handles file storage for persistent task tracking.

## Example Usage
```java
todo Finish the project
deadline Submit assignment /by 2025-01-30 23:59
event Team meeting /from 2025-01-21 3:00 PM /to 2025-01-21 4:00 PM
```

## Task List Progress
- [x] Implement core task management system
- [x] Add persistent storage for tasks
- [ ] Implement GUI interface (coming soon!)

## Motivation 💡
> "Code at night, dream in Java!" – NightCoder is designed to be your **coding companion** when burning the midnight oil.

## References
- Read more about [GitHub Flavored Markdown](https://github.github.com/gfm/)
- Inline code example: `task.setCompleted(true);`

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Shamanbenny committed:
Renamed Duke to NightCoder

Shamanbenny committed:
Updated Welcome and Exit message

Shamanbenny committed:
Updated with Echoing of User Input

Shamanbenny committed:
Updated README.md

Shamanbenny committed:
Added 'help', 'add', and 'list' command

Shamanbenny committed:
Fixed Casing Faults in previous version

Shamanbenny committed:
Updated version number for JavaDoc

Shamanbenny committed:
Added Task Object, and Done Markers

Shamanbenny committed:
Updated with 'mark' and 'unmark' command

Shamanbenny committed:
Updated with additional edge-cases & updated JavaDoc

Shamanbenny committed:
Added the inheritance of Task.java, but still unused

Shamanbenny committed:
Added 'todo', 'deadline' and 'event' command

Shamanbenny committed:
Remove all usage of Emojis in preparation for Automated Testing

Shamanbenny committed:
Updated EXPECTED.TXT and fixed Scanner bug

Shamanbenny committed:
Fixed Scanner bug and updated runtest.sh

Shamanbenny committed:
Updated EXPECTED.TXT with necessary whitespaces

Shamanbenny committed:
Updated JavaDoc version number, and checked that test passed for Windows & UNIX

Shamanbenny committed:
Added Error Handling (Exceptions) for Integer parsing during 'mark' & 'unmark'

Shamanbenny committed:
Added 'delete' feature with updated Automated Testing

Shamanbenny committed:
Used Java's Enum, Updated docs/README.md, Updated JavaDocs

Shamanbenny committed:
Update NightCoder.java with tasks file feature

Shamanbenny committed:
Add final check for data/tasks.txt after Test 2

Shamanbenny committed:
Merge Level 7 feature branch

Shamanbenny committed:
Update NightCoder with Date parsing

Shamanbenny committed:
Update text-ui-test

Shamanbenny committed:
Merge Level 8 feature branch

Shamanbenny committed:
Fix version number in class JavaDoc

Shamanbenny committed:
Add OOP for Ui, Storage, TaskList

Shamanbenny committed:
Add OOP for Parser

Shamanbenny committed:
Merge A-MoreOOP implementation

Shamanbenny committed:
Update with Packages

Shamanbenny committed:
Merge A-Packages implementation

Shamanbenny committed:
Merge Gradle support with conflict resolution

Shamanbenny committed:
Add JUnit Test for Storage.java

Shamanbenny committed:
Fix Storage.java

Shamanbenny committed:
Add JUnit Test for TaskList.java

Shamanbenny committed:
Merge A-JUnit implementation

Shamanbenny committed:
Fix build.gradle

Shamanbenny committed:
Update JavaDoc comments

Shamanbenny committed:
Apply Coding Standard

Shamanbenny committed:
Add find command implementation

Shamanbenny committed:
Update text-ui-test

Shamanbenny committed:
Merge JavaDoc updates into master

Shamanbenny committed:
Merge Coding Standard changes into master

Shamanbenny committed:
Merge Level 9 feature branch into master

hsinyilow commented on src/main/java/nightcoder/storage/Storage.java:
```diff
@@ -0,0 +1,233 @@
+     * @param completed {@code true} if the task is completed, {@code false} otherwise.
+     * @throws IOException If an I/O error occurs while updating the file.
+     */
+    public void setCompleted(int idx, boolean completed) throws IOException {
```
Maybe a different name could be used for the "completed" boolean?

Shamanbenny replied:
I agree with your suggestion. I'll be sure to push said change!

hsinyilow commented on src/main/java/nightcoder/task/Task.java:
```diff
@@ -0,0 +1,59 @@
+     */
+    @Override
+    public String toString() {
+        return (this.isCompleted ? "[X] ":"[ ] ") + this.DESCRIPTION;
```
Should the ternary expression be formatted differently?

Shamanbenny replied:
I could change it to `return ((this.isCompleted) ? "[X] " : "[ ] ") + this.DESCRIPTION;` to better fit the Coding Standard.

hsinyilow commented on src/main/java/nightcoder/parser/Parser.java:
```diff
@@ -0,0 +1,276 @@
+            this.TASKS.listTasks(findParams);
+            break;
+        default:
+            System.out.println("""
+                    \t[ Oops! ]
+                    \tI didn't catch that. Type "help" to see the list of commands I understand.
+                    \tLet's get back on track!""");
```
I like the way you split this print statement across different lines.

hsinyilow commented on src/main/java/nightcoder/parser/Parser.java:
```diff
@@ -0,0 +1,276 @@
+            // Edge-Case ['task' is already set as complete/incomplete]
+            if (completed) {
+                System.out.println("\t[ Task Already Complete! ]");
+                System.out.println("\tLooks like task \"" + task.getDescription() + "\" is already marked as done. " +
+                        "You're ahead of the game!");
```
Could consider breaking before an operator instead of after for System.out.println() lines.
Something like:
```
System.out.println("a" 
        + b);
```
instead of 
```
System.out.println("a" +
        b);
```

Shamanbenny replied:
Good catch! Indeed, according to the [Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html#layout), it is recommended to break before an operator!

hsinyilow commented on src/main/java/nightcoder/parser/Parser.java:
```diff
@@ -0,0 +1,276 @@
+        String command = parts[0].toLowerCase();
+        // The switch structure is retained for clarity and readability. Each case acts as a self-contained command
+        // handler, making the code easy to navigate without needing separate function calls for every command.
+        switch (command) {
+        case "help":
+            Ui.printHelp();
+            break;
+        case "todo":
+            if (parts.length != 2) {
+                Ui.printInvalidUsage("todo");
+                break;
+            }
+            String todoParams = parts[1];
+            this.TASKS.addToDo(todoParams);
+            break;
```
Good awareness on the indentation rules for switch case throughout.

Shamanbenny committed:
Fix Coding Standard based on PR review by hsinyilow

Shamanbenny committed:
Add checkstyle.xml

Shamanbenny committed:
Merge CheckStyle into master

Shamanbenny committed:
Update with JavaFX part 1

Shamanbenny committed:
Update with JavaFX part 2

Shamanbenny committed:
Update with JavaFX part 3

Shamanbenny committed:
Update with fully working JavaFX

Shamanbenny committed:
Merge Level-10 into master

Shamanbenny committed:
Fix empty parameter inputs for task commands

Todo, Deadline and Event command: Reject empty parameter inputs

Having a non-empty task description or timing information improves
user clarity and prevents accidental omissions.

Enforce input validation by rejecting empty parameter inputs for
Todo, Deadline, and Event commands. Introduce Java assertions to
ensure these constraints are upheld during task creation.

This change enhances data integrity and improves user experience
by encouraging meaningful task descriptions and timing details.

Shamanbenny committed:
Fix Code Quality for naming variables

Instance level final variables have been incorrectly identified as
constant variables.

Following Google's Java Style Guide, merely intending to never mutate
the object is not enough to be considered a constant variable.

Let's rename instance level final variables that are not of primitive
data typing using camelCase instead of SCREAMING_SNAKE_CASE.

Refer to this Google's Java Style Guide on 5.2.4 Constant names
https://google.github.io/styleguide/javaguide.html

Shamanbenny committed:
Implement Java Streams where applicable

TaskList.java: The listTasks methods use explicit loops and
StringBuilder to construct task listings. While functional, this
approach requires manual iteration and lacks the conciseness that
Java Streams provides.

Using Java Streams enhances readability and improves maintainability
by eliminating manual loops and leveraging stream operations for
transformation and filtering.

Let's refactor the listTasks methods to use IntStream to iterate
over task indices, replace StringBuilder with Collectors, and filter
using filter().

This update improves code clarity, reduces verbosity, fully utilizes
the functional programming possible with streams.

Shamanbenny committed:
Merge pull request #1 from Shamanbenny/branch-A-Assertions

Fix empty parameter inputs for task commands

Shamanbenny committed:
Merge branch 'master' into branch-A-CodeQuality

Shamanbenny committed:
Merge pull request #2 from Shamanbenny/branch-A-CodeQuality

Fix Code Quality for naming variables

Shamanbenny committed:
Merge pull request #3 from Shamanbenny/branch-A-Streams

Implement Java Streams where applicable

Shamanbenny committed:
Add gradle.yml as GitHub Workflow

Shamanbenny committed:
Update gradle.yml

Shamanbenny committed:
Implement detecting of duplicate tasks

Shamanbenny committed:
Merge Detect Duplicates feature branch into master

Shamanbenny committed:
Add toggle light mode and clear chat feature

Shamanbenny committed:
Merge Better Gui feature into master

Shamanbenny committed:
Update personality of NightCoder bot

Shamanbenny committed:
Update Code Quality

Shamanbenny committed:
Add Ui.png to /docs

Shamanbenny committed:
Update README.md in /docs

Shamanbenny committed:
Update and improve User Guide

Shamanbenny committed:
Fix GUI bug by removing AtlantaFX

Shamanbenny committed:
Update Ui.png

Shamanbenny committed:
Fix GUI bug and mismatched JavaFX version

Shamanbenny committed:
Update prefWidth for FXML

Shamanbenny committed:
Update MainWindow.fxml

Shamanbenny committed:
Fix GUI implementation
